src/Microsoft.DotNet.Watcher.Core/Internal/FileWatcher/PollingFileWatcher.cs

```diff
@@ -0,0 +1,243 @@
+
+namespace Microsoft.DotNet.Watcher.Core.Internal
+{
+    internal class PollingFileWatcher : IFileSystemWatcher
+    {
+        // The minimum interval to rerun the scan
+        private static readonly TimeSpan _minRunInternal = TimeSpan.FromSeconds(.5);
+
+        private readonly DirectoryInfo _watchedDirectory;
+
+        private Dictionary<string, FileMeta> _knownEntities = new Dictionary<string, FileMeta>();
+        private Dictionary<string, FileMeta> _tempDictionary = new Dictionary<string, FileMeta>();
+        private HashSet<string> _changes = new HashSet<string>();
+
+        private Thread _pollingThread;
+        private bool _raiseEvents;
+
+        private bool _disposed;
+
+        public PollingFileWatcher(string watchedDirectory)
+        {
+            if (string.IsNullOrEmpty(watchedDirectory))
+            {
+                throw new ArgumentNullException(nameof(watchedDirectory));
+            }
+
+            _watchedDirectory = new DirectoryInfo(watchedDirectory);
+
+            _pollingThread = new Thread(new ThreadStart(PollingLoop));
+            _pollingThread.IsBackground = true;
+            _pollingThread.Name = nameof(PollingFileWatcher);
+            _pollingThread.Start();
+        }
+
+        public event EventHandler<string> OnFileChange;
+
+#pragma warning disable CS0067 // not used
+        public event EventHandler OnError;
+#pragma warning restore
+
+        public bool EnableRaisingEvents
+        {
+            get
+            {
+                return _raiseEvents;
+            }
+            set
+            {
+                EnsureNotDisposed();
+
+                if (value == true)
+                {
+                    CreateKnownFilesSnapshot();
+
+                    if (_pollingThread.ThreadState == System.Threading.ThreadState.Unstarted)
+                    {
+                        // Start the loop the first time events are enabled
+                        _pollingThread.Start();
+                    }
+                }
+                _raiseEvents = value;
+            }
+        }
+
+        private void PollingLoop()
+        {
+            var stopwatch = Stopwatch.StartNew();
+            stopwatch.Start();
+
+            while (!_disposed)
+            {
+                if (stopwatch.Elapsed < _minRunInternal)
+                {
+                    // Don't run too often
+                    // The min wait time here can be double
+                    // the value of the variable (FYI)
+                    Thread.Sleep(_minRunInternal);
+                }
+
+                stopwatch.Reset();
+
+                if (!_raiseEvents)
+                {
+                    continue;
+                }
+
+                CheckForChangedFiles();
+            }
+
+            stopwatch.Stop();
+        }
+
+        private void CreateKnownFilesSnapshot()
+        {
+            _knownEntities.Clear();
+
+            ForeachEntityInDirectory(_watchedDirectory, f =>
```

Comment:
Could we change this design to only look for files we care about? For instance, move the behavior into `https://github.com/aspnet/dotnet-watch/blob/dev/src/Microsoft.DotNet.Watcher.Core/Internal/ProjectWatcher.cs` and add some sort of eventual consistency mechanism to it? Enumerating a project that uses npm could be very expensive with this approach